### ISSUE TITLE
TypeScript throwing a "Type instantiation is excessively deep and possibly infinite" error

### ISSUE THREAD

**Issue by markusylisiurunen** [runtypes/runtypes#212]
Hi!

While trying to update `runtypes` from version `5.1.0` to `5.2.0`, we encountered an issue with a fairly straightforward type definition and TypeScript throwing a `Type instantiation is excessively deep and possibly infinite` compilation error. The issue seems to only happen when first compiling a TypeScript package and then using the compiled package in a TypeScript project (so not if the type is used within the same project before compiling). We are using Yarn Workspaces so I created a minimal example where this issue happens.

You can find the example here: [https://github.com/markusylisiurunen/runtypes-excessive-depth-error](https://github.com/markusylisiurunen/runtypes-excessive-depth-error)

Here's a quick way to test:

```sh
git clone git@github.com:markusylisiurunen/runtypes-excessive-depth-error.git
cd runtypes-excessive-depth-error

yarn && cd packages/types && yarn build && cd ../client && yarn build
```

I think the example repository is the simplest way to describe the issue. If you need more details, let me know!

**Comment by markusylisiurunen** [comment]
I tried to do some digging but couldn't find a fix (at least yet). However, I managed to pinpoint the issue down to the changes made to the `Record` type between the versions `5.1.0` and `5.2.0`. It seems to have something to do with the conditional type for `FilterOptionalKeys` and `FilterRequiredKeys`.

So, this is what the types look at v5.2.0

```ts
type FilterOptionalKeys<T> = Exclude<
  {
    [K in keyof T]: T[K] extends Optional<any> ? K : never;
  }[keyof T],
  undefined
>;

type FilterRequiredKeys<T> = Exclude<
  {
    [K in keyof T]: T[K] extends Optional<any> ? never : K;
  }[keyof T],
  undefined
>;
```

and here they are after my changes

```ts
type FilterOptionalKeys<T> = Exclude<
  {
    [K in keyof T]: K;
  }[keyof T],
  undefined
>;

type FilterRequiredKeys<T> = Exclude<
  {
    [K in keyof T]: K;
  }[keyof T],
  undefined
>;
```

I know that the changes break the optional and that this is not a feasible solution. However, by removing the conditional type on `Optional<any>` the errors went away from my toy example and for our real project as well. So, definitely seems to have something to do with this.

**Comment by yuhr** [comment]
Thanks for sending the bug report, your repro is very helpful. I think there might hardly be any way around this though, I'm going to tinker around the static typing of `Record`.

By the way I have no idea how we can integrate some tests for such type of bugs, it's pretty hard to detect...

**Comment by 72636c** [comment]
Anecdotally we seem to have been hit by this once upgrading to `6.0.0`; likely arbitrary length unions and the like contribute another layer of depth.

**Comment by yuhr** [comment]
Hmm, I think I found a way. Changing the relevant code not to using `Exclude` may resolve the issue, at least in the case of the provided repro.

```ts
type FilterOptionalKeys<T> = {
  [K in keyof T]: T[K] extends Optional<any> ? K : never;
}[keyof T];

type FilterRequiredKeys<T> = {
  [K in keyof T]: T[K] extends Optional<any> ? never : K;
}[keyof T];
```

Actually I'm not sure why it originally used `Exclude` here 😕

**Comment by markusylisiurunen** [comment]
@yuhr I actually tried doing the same thing and it seemed to solve the issue for the repro but not for our real project with more complex types. I updated the repro repo with a little bit more depth to the types and the same error keeps popping up. So, unfortunately this change does not solve the issue 🙁

You can just pull the `master` from the repro repo and try building again with the version without the `Exclude`. It should throw the same error.

**Comment by yuhr** [comment]
Yeah indeed... This feature https://github.com/microsoft/TypeScript/issues/41406#issuecomment-749552535 could resolve the issue, but it's still not a complete solution (and unlikely to be implemented in the near future). We must carefully write conditional types and try to reduce their complexity as much as possible...

**Comment by paynecodes** [comment]
I know this has been acknowledged and a PR has been merged, but I wanted to throw in another example in case it helps. I have some that appears fairly simple, but consumers of these types receive the error: 

> Type instantiation is excessively deep and possibly infinite

```ts
const IProjectConfig = Record({
  packageName: String,
  projectFolder: String,
  reviewCategory: Optional(String),
  cyclicDependencyProjects: Optional(Array(String)),
  shouldPublish: Optional(Boolean),
  skipRushCheck: Optional(Boolean),
  versionPolicyName: Optional(Boolean),
});

const IRushConfig = Record({
  rushVersion: String,
  projects: Array(IProjectConfig),
  npmVersion: Optional(String),
  yarnVersion: Optional(String),
  pnpmVersion: Optional(String),
});
```

**Comment by yuhr** [comment]
@paynecodes #225 should work for your case, please check it in v6.1.0.

**Comment by markusylisiurunen** [comment]
This issue might be fixed by a PR that was just merged for TypeScript [here](https://github.com/microsoft/TypeScript/pull/45025). Looks like it might be included in the 4.5 release.

**Comment by yuhr** [comment]
This workaround https://github.com/microsoft/TypeScript/issues/34933#issuecomment-889570502 would also do the job for us.

**Comment by purge** [comment]
FYI: I get this moving from `6.4.0 -> 6.4.1` (and i've had it sporadically in previous versions)

**Comment by tshelburne** [comment]
I use runtypes to manage versions of data from backend tools and migrating between them. I'm using `Record(...).extend` to make the changes from version to version clear.

As of latest master, I'm getting "Type instantiation is excessively deep and possibly infinite.(2589)" on the last line of the following (also in [Typescript Play](https://www.typescriptlang.org/play?ts=4.9.5#code/JYWwDg9gTgLgBAKjgQwM5wEoBU4DMoQhwDkUArgHYwCeYApqsQFBO6UDGMwEFcMDMADxYAfAAoAbgC5MWAHQZKNesJEBKOAG84AXxbseqeFlp0AjHAC8shXQNQAJmM061TfkbEn6Zt0wMURnDedABMVsGmZnJ0AB78FE4ubh4wXqahfgFBIQDMESGhMfF0ic6u7gLp9LlZhsamACwFprnFCUkVqdV0jXWBDfQArC30je2lnSlVIUP9OaYAbKN0QxNlyZWeIYvzg3QA7CuL61NbaSEHe5H0ABwrB6fl09umt9chAJwrt0+b3V8PlEAAwrT5-LozEFAnwWawhMzAiEvC5RXxAA)).

```ts
import * as RT from 'runtypes'

function test<T>(v: RT.Runtype<T>) { }

const Type1 = RT.Record({})
test(Type1)

const Type2 = Type1.extend({})
test(Type2)

const Type3 = Type2.extend({})
test(Type3)

const Type4 = Type3.extend({})
test(Type4)

const Type5 = Type4.extend({})
test(Type5)

const Type6 = Type5.extend({})
test(Type6)

const Type7 = Type6.extend({})
test(Type7)

const Type8 = Type7.extend({})
test(Type8)

const Type9 = Type8.extend({})
test(Type9) // Actually throws error first here in Typescript Play

const Type10 = Type9.extend({})
test(Type10)

const Type11 = Type10.extend({})
test(Type11) // "Type instantiation is excessively deep and possibly infinite.(2589)"
```

**Comment by yuhr** [comment]
With the large refactor for v7, this should be quite less likely to happen. If someone encounters this error, please open a new issue for each case with reproductive example code.
